server/nri-api.go

```diff
@@ -0,0 +1,741 @@
+	}
+
+	a.cri.UpdateContainerLinuxResources(ctr, resources)
+
+	return nil
+}
+
+func (a *nriAPI) EvictContainer(ctx context.Context, e *api.ContainerEviction) error {
+	ctr, err := a.cri.GetContainerFromShortID(todoMissingContext(), e.ContainerId)
+	if err != nil {
+		// We blindly assume container with given ID not found and ignore it.
+		log.Errorf(ctx, "Failed to evict CRI container %q: %v", e.ContainerId, err)
+		return nil
+	}
+	err = a.cri.stopContainer(ctx, ctr, 0)
+	if err != nil {
+		log.Errorf(ctx, "Failed to evict CRI container %q: %v", e.ContainerId, err)
+		return err
+	}
+
+	return nil
+}
+
+func todoMissingContext() context.Context {
+	// TODO(klihub): call sites of this should receive the context as an
+	// argument and use that instead (typically for StartSpan()/End())
+	return context.Background()
+}
+
+//
+// NRI integration wrapper for CRI Pods
+//
+
+type criPodSandbox struct {
+	*sandbox.Sandbox
+	spec *rspec.Spec
+	pid  int
+}
+
+func nriPodSandbox(pod *sandbox.Sandbox) *criPodSandbox {
+	criPod := &criPodSandbox{
+		Sandbox: pod,
+		spec:    &rspec.Spec{},
+	}
+
+	if ic := pod.InfraContainer(); ic != nil {
+		spec := ic.Spec()
+		pid, err := ic.Pid()
+		if err != nil {
+			log.Warnf(context.TODO(), "Failed to get pid for pod infra container: %v", err)
+		}
+		criPod.spec = &spec
+		criPod.pid = pid
+	}
+
+	return criPod
+}
+
+func (p *criPodSandbox) GetDomain() string {
+	return nriDomain
+}
+
+func (p *criPodSandbox) GetID() string {
+	if p.Sandbox == nil {
+		return ""
+	}
+	return p.ID()
+}
+
+func (p *criPodSandbox) GetName() string {
+	if p.Sandbox == nil {
+		return ""
+	}
+	return p.Metadata().Name
+}
+
+func (p *criPodSandbox) GetUID() string {
+	if p.Sandbox == nil {
+		return ""
+	}
+	return p.Metadata().GetUid()
+}
+
+func (p *criPodSandbox) GetNamespace() string {
+	if p.Sandbox == nil {
+		return ""
+	}
+	return p.Metadata().Namespace
+}
+
+func (p *criPodSandbox) GetAnnotations() map[string]string {
+	if p.Sandbox == nil {
+		return nil
+	}
+	annotations := map[string]string{}
+	for key, value := range p.Annotations() {
+		annotations[key] = value
+	}
+	return annotations
+}
+
+func (p *criPodSandbox) GetLabels() map[string]string {
+	if p.Sandbox == nil {
+		return nil
+	}
+	labels := map[string]string{}
+	for key, value := range p.Labels() {
+		labels[key] = value
+	}
+	return labels
+}
+
+func (p *criPodSandbox) GetRuntimeHandler() string {
+	if p.Sandbox == nil {
+		return ""
+	}
+	return p.RuntimeHandler()
+}
+
+func (p *criPodSandbox) GetLinuxPodSandbox() nri.LinuxPodSandbox {
+	return p
+}
+
+func (p *criPodSandbox) GetLinuxNamespaces() []*api.LinuxNamespace {
+	if p.spec.Linux == nil {
+		return nil
+	}
+	return api.FromOCILinuxNamespaces(p.spec.Linux.Namespaces)
+}
+
+func (p *criPodSandbox) GetPodLinuxOverhead() *api.LinuxResources {
+	if p.Sandbox == nil {
+		return nil
+	}
+	// TODO(klihub): AFAICT, CRI PodSandboxConfig is not stored, so we can only
```

Comment:
is this TODO expected to not be resolved? can we open an issue for it if this merges?

Reply:
Yes, I don't expect those to be resolved now, that's why I left the TODOs in place. I can open an issue for this post-merge (and can also take a shot at implementing the missing bits).